src/core/layout/qgslayoutitemattributetable.cpp

```diff
@@ -575,9 +576,27 @@ bool QgsLayoutItemAttributeTable::getTableContents( QgsLayoutTableContents &cont
 
         if ( mUseConditionalStyling )
         {
+          QgsConditionalStyle constraintstyle;
+          if ( QgsVectorLayerUtils::attributeHasConstraints( layer, idx ) )
```

Comment:
Hmm -- I'm not sure we should always apply these here. From a user perspective conditional styles are a different thing to attribute constraints.

Reply:
are you suggesting that the conditional styling for constraints should only be shown on demand?

Reply:
would that mitigate your concerns about general performance?

Reply:
> are you suggesting that the conditional styling for constraints should only be shown on demand?

Yes, definitely. We just can't afford any additional slowness in the attribute table.

But my concern is that in practice this will be unusably slow for any moderate sized layer (with a unique constraint, which the majority do), even if it's opt in. We've got:

- in the filter each feature will trigger a complete table traversal, so n^2 feature retrievals
- an additional n check of the constraint when evaluating the conditional style, so another n^2 retrievals

I think correct approach would be to "expand" the constraints to a combined qgis expression, so that this can be then compiled when preparing the feature request. That would help at lot with expression based constraints. We'd also need some new function for testing uniqueness of field values, and then would need to compile that too....